File: docs/vectors/dpop/rfc9449/verify_vectors.mjs
#!/usr/bin/env node
/** Dependency-free Node.js verification of the RFC 9449 resource-server vectors. */

import { createHash, createPublicKey, verify } from "node:crypto";
import { readFile } from "node:fs/promises";

const manifestUrl = new URL("./manifest.json", import.meta.url);
const data = JSON.parse(await readFile(manifestUrl, "utf8"));
const now = Date.parse(data.now) / 1000;

function decodePart(part) {
  return JSON.parse(Buffer.from(part, "base64url").toString("utf8"));
}

function decodeJwt(value) {
  const parts = value.split(".");
  if (parts.length !== 3) throw new Error("malformed JWT");
  return {
    header: decodePart(parts[0]),
    claims: decodePart(parts[1]),
    signingInput: Buffer.from(`${parts[0]}.${parts[1]}`),
    signature: Buffer.from(parts[2], "base64url"),
  };
}

function verifyEs256(jwt, jwk) {
  return verify(
    "sha256",
    jwt.signingInput,
    { key: createPublicKey({ key: jwk, format: "jwk" }), dsaEncoding: "ieee-p1363" },
    jwt.signature,
  );
}

function thumbprint(jwk) {
  const canonical = JSON.stringify({ crv: jwk.crv, kty: jwk.kty, x: jwk.x, y: jwk.y });
  return createHash("sha256").update(canonical).digest("base64url");
}

function normalizeHtu(value) {
  const url = new URL(value);
  if (url.protocol !== "https:" || url.username || url.password) throw new Error("untrusted htu");
  url.search = "";
  url.hash = "";
  if (url.port === "443") url.port = "";
  return url.toString().replace(/\/$/, url.pathname === "/" ? "/" : "");
}

function classify(testCase) {
  if (testCase.authorization?.startsWith("Bearer ") && testCase.dpop) return "ambiguous";
  if (!testCase.authorization?.startsWith("DPoP ") || !testCase.dpop) return "malformed";
  try {
    const accessTokenValue = testCase.authorization.slice(5);
    const accessToken = decodeJwt(accessTokenValue);
    const proof = decodeJwt(testCase.dpop);
    const issuerJwk = data.jwks.keys.find((key) => key.kid === accessToken.header.kid);
    if (
      accessToken.header.typ !== "at+jwt" ||
      accessToken.header.alg !== "ES256" ||
      !issuerJwk ||
      !verifyEs256(accessToken, issuerJwk)
    ) return "invalid";
    if (
      accessToken.claims.iss !== data.issuer ||
      accessToken.claims.aud !== data.audience ||
      accessToken.claims.iat > now + 30 ||
      accessToken.claims.exp <= now - 30
    ) return "invalid";
    if (
      proof.header.typ !== "dpop+jwt" ||
      proof.header.alg !== "ES256" ||
      proof.header.jwk?.d ||
      !verifyEs256(proof, proof.header.jwk)
    ) return "invalid";
    if (
      proof.claims.htm.toUpperCase() !== testCase.method.toUpperCase() ||
      normalizeHtu(proof.claims.htu) !== normalizeHtu(testCase.target_uri) ||
      Math.abs(now - proof.claims.iat) > 90
    ) return "sender_constraint_mismatch";
    const ath = createHash("sha256").update(accessTokenValue, "ascii").digest("base64url");
    if (proof.claims.ath !== ath || accessToken.claims.cnf?.jkt !== thumbprint(proof.header.jwk)) {
      return "sender_constraint_mismatch";
    }
    return "pass";
  } catch {
    return "invalid";
  }
}

let failures = 0;
for (const testCase of data.cases) {
  const actual = classify(testCase);
  const expected = testCase.expect === "pass" ? "pass" : testCase.expect === "ambiguous" ? "ambiguous" : testCase.failure;
  if (actual !== expected) {
    console.error(`FAIL ${testCase.id}: expected ${expected}, got ${actual}`);
    failures += 1;
  } else {
    console.log(`ok  ${testCase.id}`);
  }
}
process.exitCode = failures === 0 ? 0 : 1;

File: docs/vectors/dpop/rfc9449/verify_vectors.py
#!/usr/bin/env python3
"""Verify language-neutral DPoP RS vectors against authweave-workload[dpop]."""
# ruff: file-ignore[print]

from __future__ import annotations

import asyncio
import json
from datetime import datetime
from pathlib import Path

from authweave_core import (
    Authenticated,
    AuthenticationRuntime,
    CredentialMatch,
    FailureCode,
    InMemoryReplayStore,
    Invalid,
    RequestView,
)
from authweave_workload.dpop import DPoPBoundJWTProvider, DPoPPolicy
from authweave_workload.jwks import BoundedJWKSClient
from authweave_workload.jwt import TrustedIssuer

ROOT = Path(__file__).resolve().parent
MANIFEST = ROOT / "manifest.json"


class _Clock:
    def __call__(self) -> float:
        return 0.0


async def _run() -> int:
    data = json.loads(MANIFEST.read_text())
    now = datetime.fromisoformat(data["now"])
    failures = 0
    for case in data["cases"]:
        provider = DPoPBoundJWTProvider(
            name="vector_rs",
            issuer=TrustedIssuer(
                issuer=data["issuer"],
                audiences=frozenset({data["audience"]}),
                environment="sandbox",
                jwks=BoundedJWKSClient(static_jwks=data["jwks"]),
            ),
            dpop=DPoPPolicy(resource_server_id=data["resource_server_id"]),
            replay_store=InMemoryReplayStore(capacity=64, time_source=_Clock()),
        )
        headers: list[tuple[bytes, bytes]] = []
        if case.get("authorization"):
            headers.append((b"authorization", case["authorization"].encode()))
        if case.get("dpop"):
            headers.append((b"dpop", case["dpop"].encode()))
        request = RequestView(
            method=case["method"],
            headers=tuple(headers),
            timestamp=now,
            target_uri=case["target_uri"],
        )
        match = provider.match(request)
        if case["expect"] == "ambiguous":
            ok = match is CredentialMatch.AMBIGUOUS
            if not ok:
                print(f"FAIL {case['id']}: match={match}")
                failures += 1
            else:
                print(f"ok  {case['id']}")
            continue
        if match is not CredentialMatch.OWNED:
            print(f"FAIL {case['id']}: expected OWNED got {match}")
            failures += 1
            continue
        decision = await provider.authenticate(request, AuthenticationRuntime())
        if case["expect"] == "pass":
            ok = isinstance(decision, Authenticated)
        else:
            expected = FailureCode(case["failure"])
            ok = isinstance(decision, Invalid) and decision.code is expected
        if not ok:
            print(f"FAIL {case['id']}: decision={decision!r}")
            failures += 1
        else:
            print(f"ok  {case['id']}")
    return failures


if __name__ == "__main__":
    raise SystemExit(asyncio.run(_run()))
